dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+                }
+            }
+            else
+            {
+                Value(parsedNum.Value());
+            }
+        }
+    }
+}
+
+void NumberBox::OnSpinDownClick(winrt::IInspectable const&  sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(false);
+}
+
+void NumberBox::OnSpinUpClick(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(true);
+}
+
+void NumberBox::OnNumberBoxKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+    switch (args.OriginalKey())
+    {
+        case winrt::VirtualKey::Enter:
+        case winrt::VirtualKey::GamepadA:
+            ValidateInput();
+            break;
+
+        case winrt::VirtualKey::Escape:
+        case winrt::VirtualKey::GamepadB:
+            UpdateTextToValue();
+            break;
+
+        case winrt::VirtualKey::Up:
+            StepValue(true);
+            break;
+
+        case winrt::VirtualKey::Down:
+            StepValue(false);
+            break;
+    }
+}
+
+void NumberBox::OnScroll(winrt::IInspectable const& sender, winrt::PointerRoutedEventArgs const& args)
+{
+    if (HyperScrollEnabled())
+    {
+        int delta = args.GetCurrentPoint(*this).Properties().MouseWheelDelta();
+        if (delta > 0)
+        {
+            StepValue(true);
+        }
+        else if (delta < 0)
+        {
+            StepValue(false);
+        }
+    }
+}
+
+void NumberBox::StepValue(bool isPositive)
+{
+    double oldVal = Value();
+    double newVal = oldVal;
+
+    if (isPositive)
+    {
+        newVal += StepFrequency();
+    }
+    else
+    {
+        newVal -= StepFrequency();
+    }
+
+    if (WrapEnabled())
+    {
+        if (newVal > Maximum())
+        {
+            newVal = Minimum();
+        }
+        else if (newVal < Minimum())
+        {
+            newVal = Maximum();
+        }
+    }
+
+    // Safeguard for floating point imprecision errors
+    int StepFreqSigDigits = ComputePrecisionRounderSigDigits(newVal);
+    m_stepPrecisionRounder.SignificantDigits(StepFreqSigDigits);
+    newVal = m_stepPrecisionRounder.RoundDouble(newVal);
+
+    // Update Text and Revalidate new value
+    Value(newVal);
+}
+
+// Check if text resembles formulaic input to determine if parser should be executed
+bool NumberBox::IsFormulaic(const winrt::hstring& in)
+{
+    std::wstring input(in);
+    std::wregex formula(L"^([0-9()\\s]*[+-/*^%]+[0-9()\\s]*)+$");
+    std::wregex negval(L"^-([0-9])+(.([0-9])+)?$");
+    return (std::regex_match(input, formula) && !std::regex_match(input, negval));
+}
+
+// Computes the number of significant digits that precision rounder should use. This helps to prevent floating point imprecision errors. 
+int NumberBox::ComputePrecisionRounderSigDigits(double newVal)
+{
+    auto const oldVal = Value();
+
+    // Run formatter on both values to discard trailing and leading 0's.
+    std::wstring formattedVal(m_stepPrecisionFormatter.Format(oldVal));
```

Comment:
>std::wstring [](start = 4, length = 12)

prefer auto to avoid a full copy of these strings to a new heap allocation. auto will receive this as a winrt::hstring (just adds a ref, not copied). 

hstring might not have those convenience methods like find_first_of but it converts into wstring_view without an allocation, and wstring_view should have those.

Reply:
Did this work?

---
In reply to: [340957147](https://github.com/microsoft/microsoft-ui-xaml/pull/1507#discussion_r340957147) [](ancestors = 340957147)